# Stage A Quiz

In [1]:
import pandas as pd
import numpy as np

In [2]:
food = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding='latin-1')

In [3]:
food.shape

(60943, 12)

In [4]:
food.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

In [5]:
food.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

# Q1. What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [6]:
num_miss = food.isnull().sum()
num_miss

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

**Ans: 1589 and 2.607%**

In [7]:
perc_miss = num_miss/sum(num_miss)
round(perc_miss * 100, 2)

Area Code        0.00
Area             0.00
Item Code        0.00
Item             0.00
Element Code     0.00
Element          0.00
Unit             0.00
Y2014           20.87
Y2015           20.33
Y2016           20.16
Y2017           19.78
Y2018           18.86
dtype: float64

# Q2. Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [8]:
element_process = food.groupby('Element').sum()
element_process

C:\Users\USER\AppData\Local\Temp\ipykernel_96656\341399250.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  element_process = food.groupby('Element').sum()


Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

**Ans: 292836.00**

# Q3. Which year had the least correlation with ‘Element Code’?

In [9]:
food.corr()

C:\Users\USER\AppData\Local\Temp\ipykernel_96656\1888934596.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  food.corr()


Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

**Ans: Y2016**

# Q4. Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

In [10]:
element = food.groupby('Element').sum()
element

C:\Users\USER\AppData\Local\Temp\ipykernel_96656\2277617435.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  element = food.groupby('Element').sum()


Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

**Ans: 2014**

# Q5. What is the total number of unique countries in the dataset?

In [11]:
len(food.Area.unique())

49

# Q6. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [12]:
food.groupby('Area')['Y2017'].sum().sort_values()

Area
Sudan (former)                       0.00
Ethiopia PDR                         0.00
Comoros                             59.84
Seychelles                         442.34
Sao Tome and Principe            12662.63
Cabo Verde                       14650.74
Guinea-Bissau                    19102.77
Lesotho                          21267.96
Botswana                         22101.30
Djibouti                         22729.91
Gambia                           23154.18
Gabon                            27979.64
Liberia                          29342.20
Namibia                          29874.89
Central African Republic         29937.00
Congo                            41181.68
Togo                             49841.88
Mauritius                        51114.83
Eswatini                         54343.33
Sierra Leone                     55311.33
Chad                             71594.68
Rwanda                           73663.69
Zimbabwe                         75919.34
Senegal                      

**Ans: Guinea-Bissau**

# Q7. What is the total Protein supply quantity in Madagascar in 2015?

In [13]:
protein = food[food.Area == 'Madagascar']
protein_2015 = protein.groupby('Element').sum()
protein_2015

C:\Users\USER\AppData\Local\Temp\ipykernel_96656\531382540.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  protein_2015 = protein.groupby('Element').sum()


Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                    15351     318888        630819   
Export Quantity                             13416     279445        614744   
Fat supply quantity (g/capita/day)          14319     299405         75924   
Feed                                         4515      94030        193235   
Food                                        14190     295817        565620   
Food supply (kcal/capita/day)               14319     299405         73704   
Food supply quantity (kg/capita/yr)         14190     295817         70950   
Import Quantity                             14319     297945        622821   
Losses                                       7095     148102        281765   
Other uses (non-food)                        6579     138246        262854   
Processing                                   9804     201929        389956   
Production                                  14190     295071        606210   
Protein supply quantity (g/capita/day)      14319     299405         74814   
Residuals                                   13545     280011        542850   
Seed                                         2709      56067        116067   
Stock Variation                             12642     262174        497056   
Total Population - Both sexes                 129       2501           511   

                                           Y2014     Y2015     Y2016  \
Element                                                                
Domestic supply quantity                31578.76  31214.98  31732.08   
Export Quantity                           425.48    494.75    535.06   
Fat supply quantity (g/capita/day)         88.72     91.85     95.33   
Feed                                     2178.76   2070.22   2121.40   
Food                                    21405.42  21120.65  21435.61   
Food supply (kcal/capita/day)            7945.00   7685.00   7669.00   
Food supply quantity (kg/capita/yr)       907.59    871.59    861.17   
Import Quantity                          2010.71   1721.80   1901.47   
Losses                                   2252.00   2176.00   2215.00   
Other uses (non-food)                    4348.58   3957.12   4114.08   
Processing                               1262.00   1803.00   1770.00   
Production                              30337.51  29482.89  29889.65   
Protein supply quantity (g/capita/day)    177.79    173.05    173.40   
Residuals                                -603.00   -616.00   -642.00   
Seed                                      728.00    699.00    720.00   
Stock Variation                           352.00   -505.00   -475.00   
Total Population - Both sexes           23590.00  24234.00  24894.00   

                                           Y2017     Y2018  
Element                                                     
Domestic supply quantity                31927.87  32596.87  
Export Quantity                           566.81    535.81  
Fat supply quantity (g/capita/day)        101.03     96.43  
Feed                                     2059.60   2133.60  
Food                                    21676.21  22116.21  
Food supply (kcal/capita/day)            7768.00   7748.00  
Food supply quantity (kg/capita/yr)       847.75    842.17  
Import Quantity                          3139.79   3214.79  
Losses                                   2153.00   2253.00  
Other uses (non-food)                    4170.06   4221.06  
Processing                               1826.00   1819.00  
Production                              29267.86  30215.86  
Protein supply quantity (g/capita/day)    175.75    176.64  
Residuals                                -657.00   -676.00  
Seed                                      695.00    728.00  
Stock Variation                           -91.00    293.00  
Total Population - Both sexes           25571.00  26262.00

**Ans: 173.05**

# Q8. Answer the following questions based on the African food production dataset provided by the FAO website already provided. What is the total sum of Wine produced in 2015 and 2018 respectively?

In [14]:
wine_prod = food.groupby('Item').sum()
wine_prod

C:\Users\USER\AppData\Local\Temp\ipykernel_96656\3088225372.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  wine_prod = food.groupby('Item').sum()


Area Code  Item Code  Element Code      Y2014      Y2015  \
Item                                                                            
Alcohol, Non-Food        38447     752497       1515090    2403.00    2180.00   
Alcoholic Beverages      75016    1649136       2171381  102410.11   98783.72   
Animal Products          18060     397035         90990   11935.65   11811.73   
Animal fats              85279    1882494       2567303  209460.54  200675.72   
Apples and products      69280    1352989       1920418    9499.23   10559.15   
...                        ...        ...           ...        ...        ...   
Vegetables, Other        77779    1518715       2275409  155038.96  158104.08   
Vegetal Products         18060     391905         90990  107145.19  107064.17   
Wheat and products       87073    1639683       2654934  232670.13  234710.51   
Wine                     66206    1319535       1818328    4497.36    4251.81   
Yams                     50798     970905       1525745  200396.96  203151.78   

                         Y2016      Y2017      Y2018  
Item                                                  
Alcohol, Non-Food      2223.00    2348.00    2293.00  
Alcoholic Beverages   96958.75   95581.06   97847.27  
Animal Products       11661.69   11547.65   11578.61  
Animal fats          183314.22  269617.53  269648.27  
Apples and products    9853.26   10198.90    9640.51  
...                        ...        ...        ...  
Vegetables, Other    156218.90  157752.59  163987.21  
Vegetal Products     106792.72  107655.20  107775.39  
Wheat and products   239377.07  240047.62  242645.19  
Wine                   3872.09    4178.02    4039.32  
Yams                 220626.20  229174.59  221272.09  

[119 rows x 8 columns]

**Ans: 4251.81 and 4039.32**

# Q9. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [15]:
food.groupby('Area')['Y2017'].sum().sort_values(ascending=False)

Area
Nigeria                        1483268.23
Egypt                           866379.92
South Africa                    517590.54
Ethiopia                        448683.76
Morocco                         388495.36
Ghana                           337599.06
Algeria                         325644.27
United Republic of Tanzania     322616.85
Kenya                           264660.66
Sudan                           239931.92
Cameroon                        232030.43
Angola                          229159.57
Côte d'Ivoire                   224599.01
Uganda                          213950.38
Malawi                          181098.71
Mozambique                      161407.98
Mauritania                      156665.46
Mali                            149928.33
Madagascar                      131197.73
Niger                           126707.58
Benin                           124771.22
Tunisia                         124167.20
Zambia                          103223.77
Burkina Faso                 

**Ans: Nigeria**

# Q10 What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [16]:
food.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

**Ans: 140.92 and 1671.86**